metrics/serve.go

```diff
@@ -94,30 +86,49 @@ func RunMetrics(metricCount int, labelCount int, seriesCount int, metricLength i
 	valueTick := time.NewTicker(time.Duration(valueInterval) * time.Second)
 	seriesTick := time.NewTicker(time.Duration(seriesInterval) * time.Second)
 	metricTick := time.NewTicker(time.Duration(metricInterval) * time.Second)
+	updateNotify := make(chan struct{}, 1)
 
 	go func() {
 		for tick := range valueTick.C {
 			fmt.Printf("%v: refreshing metric values\n", tick)
+			metricsMux.Lock()
 			cycleValues(labelKeys, labelValues, seriesCount, seriesCycle)
+			metricsMux.Unlock()
+			select {
+			case updateNotify <- struct{}{}:
+			default:
+			}
 		}
 	}()
 
 	go func() {
 		for tick := range seriesTick.C {
 			fmt.Printf("%v: refreshing series cycle\n", tick)
+			metricsMux.Lock()
```

Comment:
moved the lock here because otherwise there is a race for `seriesCycle++` causing `cycleValues` to misbehave